New input: GridMap (canvas texture, etc)

briend opened the pull request:
This input allows interesting surface textures like canvas, laid paper, etc.  It can also create very dynamic brushes based on an underlying grid-like pattern.  The important difference between this and Stroke is that any patterns for Stroke will follow the path of the brush, whereas patterns on GridMap stick to the canvas.  More info at the community forum:
https://community.mypaint.org/t/new-input-surfacemap-canvas-texture-etc/473

briend commented:
I switched the scale to logarithmic which makes a lot more sense (I think) for scaling it with zoom-level and brush-size (brush size is already logarithmic)

achadwick commented:
This is really fun. By the way, I had the same misunderstanding as [Rebecca in the forums](https://community.mypaint.org/t/new-input-surfacemap-canvas-texture-etc/473/9) - for some reason I thought a mask would be involved, but in practice you have to use it with a brush whose radius is very much smaller than the peaks you pass in through this feature's inputs and settings in order to get anything.

* Can we change the name to something other than "Surface" map, since no surface is involved as input to the mapping? I want there to be a way of feeding an alpha mask to the brush engine at some unspecified future date so that dabs are masked. 

* Should Scale, Scale X, and Scale Y be subject to brush dynamics at all? It seems like it would be better to have them as constants.

* Please change the "tooltip" texts: Scale X and Scale Y are not percentages. Also, it could be made clearer that the settings affect the inputs ("brush dynamics") elsewhere.

achadwick commented:
The texts might be read as indicating that the effect size changes with the brush radius, which isn't the case.

![surfacemap_test](https://cloud.githubusercontent.com/assets/61299/23104591/ae5fe7bc-f6c8-11e6-80f0-36a6727783d9.png)

briend commented:
Thanks Andrew!  I think I fixed it all.  I renamed it to "GridMap" since. . .it's basically a grid... mapping values to a location.... Sound good?  I would really like to keep the settings dynamic if we can, I've already had a couple good use cases for changing the scales dynamically.  That [fractal-demo ](https://youtu.be/DGd88-xM1Rs?t=66) mapped the GridMap scale to zoom-level to make it seamless.  Also, adding some "random" to the grid scale can give a neat jitter effect to the whole pattern, softening it, etc.  Tilt can be really interesting to apply to the X and Y scales, especially if I can figure out how to isolate tilt X vs tilt Y....  well, without adding yet more inputs :-)

QuLogic commented on mypaint-brush.c:
```diff
@@ -553,8 +553,8 @@ smallest_angular_difference(float angleA, float angleB)
-    inputs[MYPAINT_BRUSH_INPUT_GRIDMAP_X] = self->states[MYPAINT_BRUSH_STATE_GRIDMAP_X];
-    inputs[MYPAINT_BRUSH_INPUT_GRIDMAP_Y] = self->states[MYPAINT_BRUSH_STATE_GRIDMAP_Y];
+    inputs[MYPAINT_BRUSH_INPUT_GRIDMAP_X] = CLAMP(self->states[MYPAINT_BRUSH_STATE_GRIDMAP_X], 0.0, 256.0);
+    inputs[MYPAINT_BRUSH_INPUT_GRIDMAP_Y] = CLAMP(self->states[MYPAINT_BRUSH_STATE_GRIDMAP_Y], 0.0, 256.0);
```
Is this supposed to fit in a byte? The max there is 255.

briend replied:
No, I just arbitrarily made the base grid size 256 pixels wide.  I'm doing a mod(X, 256) but scaling everything which seems to cause some rounding issues.

briend commented:
I made a demo with some half tone brushes here, pretty fun!
https://www.youtube.com/watch?v=g2fwdfg7VdA

briend commented:
zoomed in look at the CMY halftone printer-like brush I'm working on.
https://gfycat.com/gifs/detail/ForsakenWebbedCockroach

briend commented on tilemap.c:
```diff
@@ -29,7 +29,7 @@ tile_map_new(int size, size_t item_size, TileMapItemFreeFunc item_free_func)
     self->item_size = item_size;
     self->item_free_func = item_free_func;
-    const int map_size = 2*self->size*2*self->size;
+    const float map_size = 2*self->size*2*self->size;
```
Is float a bad idea?

0ion9 replied:
.. I hope I'm misunderstanding something here:

* Line 32 said `const int map_size = 2*self->size*2*self->size;` I'm not sure how the compiler refactors that, but this in combination with the stated error seems to imply self->size=40960, meaning 6,710,886,400 is the value assigned to this.
* Line 33 multiplies this by self->item_size - which seems as if it must be a pointer, therefore either 4 or 8 bytes. So it requests to malloc either 26,843,545,600 bytes or 53,687,091,200 bytes
* ie. either 25GB or 50GB.
* If this is true, why is it happening?

* Line 65 just confirms my hypothesis about the value of self->item_size
* Not sure what the line 71 report is about; there isn't any multiplying happening there.

Putting aside that allocation issue.. Isn't `size_t` the naturally appropriate type for passing to `malloc` (cf. `man malloc`)? IMO both `map_size` and `i` should be of this type.

briend replied:
It looks like the size is affected by HEAVY_DEBUG in operationqueue.c.  I need to verify that I didn't turn debug and that's what fixed it.  I don't think I'm suddenly allocating 25GB of ram and not noticing .. .

briend commented:
I removed the float patch, that was a total red herring.  There is definitely something causing memory growth.  I've included the mapping patch that avoids redundant interpolation and that makes it about 100 times better.  It won't try to consume 500GB of ram anymore, but it WILL slowly grow to 2GB or more just by fiddling with live update settings.

I will document more here:
https://github.com/mypaint/libmypaint/issues/99

briend commented:
> The texts might be read as indicating that the effect size changes with the brush radius, which isn't the case.
Combined with the new Brush Base Radius input, this is now possible.  The HalfTone brushes (you can find them in my master branch) I demo'ed above can now be resized on-the-fly and the scale of the gridmap changes in-tune with the brush size.  It's really pretty cool.  I'd like to merge both of these new inputs soon if there are no objections.
